generator/src/main/java/com/graphicsfuzz/generator/transformation/donation/MakeArrayAccessesInBounds.java

```diff
@@ -51,20 +52,37 @@ public void visitArrayIndexExpr(ArrayIndexExpr arrayIndexExpr) {
     type = type.getWithoutQualifiers();
     assert isArrayVectorOrMatrix(type);
     if (!staticallyInBounds(arrayIndexExpr.getIndex(), type)) {
-      arrayIndexExpr.setIndex(new TernaryExpr(
+      Type indexType = typer.lookupType(arrayIndexExpr.getIndex());
+      if (indexType != null) {
+        indexType = indexType.getWithoutQualifiers();
+      }
+      if (indexType == BasicType.UINT) {
+        arrayIndexExpr.setIndex(new TernaryExpr(
             new BinaryExpr(
-                  new BinaryExpr(
-                        new ParenExpr(arrayIndexExpr.getIndex().clone()),
-                        new IntConstantExpr("0"),
-                        BinOp.GE),
-                  new BinaryExpr(
-                        new ParenExpr(arrayIndexExpr.getIndex().clone()),
-                        new IntConstantExpr(getSize(type).toString()),
-                        BinOp.LT),
-                  BinOp.LAND),
+                new ParenExpr(arrayIndexExpr.getIndex().clone()),
+                new UIntConstantExpr(getSize(type).toString() + "u"),
+                BinOp.LT),
+            arrayIndexExpr.getIndex(),
+            new UIntConstantExpr("0u"))
+        );
+      } else {
+        // If indexType ends up being null and the expression is still of type uint, it's better to
+        // generate a statement that will fail to validate rather than crash GraphicsFuzz.
+        arrayIndexExpr.setIndex(new TernaryExpr(
```

Comment:
This is a seemingly minor problem I ran into when developing the tests:

```vec3 data[16];
uint uselessOutVariable;
vec3 stuff = data[uaddCarry(19u, 15u, uselessOutVariable)];```

If you try to make this inbounds *in a test case,* indexType will be null and so you'll hit the `else` case here. However, it seems that fuzzing shaders with expressions like these (specifically the one that was causing PR #605 to fail CI) does not cause this to happen. I am not sure why, but it is probably worth looking into a little.

Reply:
You seem to have something equivalent to that in the tests for this PR, but I could not see it making indexType turn out to be null.

Can you elaborate a bit more?  We indeed don't really want indexType to be null unless there's some good reason (which I can't think of).

Reply:
Yeah - if you put this uaddCarry call inside of the array access directly(like in the above comment), Typer is unable to find its type and so it will be null. However, making a variable and assigning the function call before using the array access with the variable (like it's done in the test) will work properly.

Reply:
The issue there turns out to be due to uaddCarry not being available in GLSL ES 3.00.  Being unknown, it has not been given a type.

I think for this function what I would prefer would be:

      Type indexType = typer.lookupType(arrayIndexExpr.getIndex());
      if (indexType == null) {
          return;
      }
      indexType = indexType.getWithoutQualifiers();

And then in the else branch an assertion that indexType is BasicType.INT.

Ultimately we would like the whole AST to be typed, but at present some things can end up being untyped.  At the moment what the method already does is returns in such cases (see the start of the method), so I think it would be good to follow that style here.

Reply:
Oops, that's kind of a silly mistake on my part. No wonder it worked fine in the shader.

All fixed.